src/lazy/encoder/binary/v1_1/value_writer.rs

```diff
@@ -231,8 +232,253 @@ impl<'value, 'top> BinaryValueWriter_1_1<'value, 'top> {
+        // The number of bits dedicated to representing the offset in a long-form timestamp
+        const NUM_OFFSET_BITS: u32 = 12;
+        // The bit offsets of each time unit within the encoding
+        const MONTH_BIT_OFFSET: u32 = NUM_YEAR_BITS;
+        const DAY_BIT_OFFSET: u32 = MONTH_BIT_OFFSET + NUM_MONTH_BITS;
+        const HOUR_BIT_OFFSET: u32 = DAY_BIT_OFFSET + NUM_DAY_BITS;
+        const MINUTE_BIT_OFFSET: u32 = HOUR_BIT_OFFSET + NUM_HOUR_BITS;
+        const OFFSET_BIT_OFFSET: u32 = MINUTE_BIT_OFFSET + NUM_MINUTE_BITS;
+        const SECOND_BIT_OFFSET: u32 = OFFSET_BIT_OFFSET + NUM_OFFSET_BITS;
+
+        let mut encoding: u64 = value.year() as u64;
+        encoding |= (value.month() as u64) << MONTH_BIT_OFFSET;
+        encoding |= (value.day() as u64) << DAY_BIT_OFFSET;
+        encoding |= (value.hour() as u64) << HOUR_BIT_OFFSET;
+        encoding |= (value.minute() as u64) << MINUTE_BIT_OFFSET;
+
+        const MIN_OFFSET: i32 = -24 * 60;
+        const UNKNOWN_OFFSET: u64 = (1 << NUM_OFFSET_BITS) - 1;
+
+        let offset_minutes = match value.offset() {
+            Some(minutes) => (minutes - MIN_OFFSET) as u64,
+            None => UNKNOWN_OFFSET,
+        };
+        encoding |= offset_minutes << OFFSET_BIT_OFFSET;
+        encoding |= (value.second() as u64) << SECOND_BIT_OFFSET;
+
+        let precision = value.precision();
+        let scale = value.fractional_seconds_scale().unwrap_or(0);
+
+        // The encoded bytelength of all components *except* subseconds.
+        let (mut encoded_length, num_bits_in_use) = match precision {
+            Year => (2, 14),
+            Month => (3, MONTH_BIT_OFFSET + NUM_MONTH_BITS),
+            Day => {
+                let bits_in_use = DAY_BIT_OFFSET + NUM_DAY_BITS;
+                // Set the "day-not-month" bit, just beyond the day bits.
+                encoding |= 1 << bits_in_use;
+                (3, bits_in_use + 1)
+            }
+            // (hour, minute, offset) are an atomic unit in the encoding--when one is present they must all be present.
+            HourAndMinute => (6, OFFSET_BIT_OFFSET + NUM_OFFSET_BITS),
+            Second => (7, SECOND_BIT_OFFSET + SECOND_BIT_OFFSET),
+        };
+
+        // Because we eagerly (and branchless-ly) encoded all of the time units, we may have populated bits that are
+        // irrelevant to the final encoding. To simplify unit testing (and in the current absence of a binary 1.1
+        // reader), we calculate a mask of which bits are relevant to the current opcode and set any bits not in use
+        // to `0`.
+        // TODO: Remove this logic pending resolution of https://github.com/amazon-ion/ion-docs/issues/295, which
+        //       suggests requiring readers to ignore bits not used by the specified opcode
+        let mask = 1u64
+            .checked_shl(num_bits_in_use)
+            .unwrap_or(0)
+            .checked_sub(1)
+            .unwrap_or(u64::MAX);
```

Comment:
FYI, this could be replaced with [`wrapping_sub`](https://doc.rust-lang.org/stable/std/primitive.u64.html#method.wrapping_sub).

Reply:
Oh yeah, good call 👌

Reply:
In the same spirit, I tried to replace `checked_shl`/`unwrap_or` with `wrapping_shl` and learned that trying to left shift by more bits than the type uses is undefined behavior (and not just `0`). Wild.